BUG: groupby.shift returns different columns when fill_value is specified

smithto1 opened the pull request:
- [x] closes #41556
- [x] closes #26615 
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

This fixes the minimal reproducing examples from the original bug report #41556. 

**With master**, specifying `fill_value` causes the index columns to be returned and to include the `fill_value` in the grouping columns. 

```
In [4]: import pandas as pd
   ...: 
   ...: df = pd.DataFrame({'a': [2, 1, 2, 1], 'b': ['x', 'x', 'y', 'y']})
   ...: 
   ...: df.groupby('a').shift(1)
Out[4]:
     b
0  NaN
1  NaN
2    x
3    x
In [5]: df.groupby('a').shift(1, fill_value='fill')
Out[5]:
      a     b
0  fill  fill
1  fill  fill
2     2     x
3     1     x
```

**With this PR**, only the value columns are returned in both cases, and the fill is applied correctly.

```
In [1]: import pandas as pd
   ...: 
   ...: # On master, using fill_value causes the index columns to be returned
   ...: df = pd.DataFrame({'a': [2, 1, 2, 1], 'b': ['x', 'x', 'y', 'y']})
   ...: 
   ...: df.groupby('a').shift(1)
Out[1]:
     b
0  NaN
1  NaN
2    x
3    x
In [2]: df.groupby('a').shift(1, fill_value='fill')
Out[2]:
      b
0  fill
1  fill
2     x
3     x
```

On master, if `fill_value=None` then `_get_cythonized_result` was used. But if `fill_value` was specified, then `self.apply` was used, because `_get_cythonized_result` couldn't take the `fill_value`. I've updated `_get_cythonized_result` so it can handle the `fill_value` itself. This means `Groupby.shift` follows the same code path and returns the same structure of output, whether or not `fill_value` is specified. 

**Test**
The bug reported in the original issue should have been caught by an existing test, except that the existing test extracts the values columns ( `[['Z']]` ), before doing its comparison to the expected output. This was hiding the fact that the structure of the result depended on `fill_value` being specified. 

https://github.com/pandas-dev/pandas/blob/1.2.x/pandas/tests/groupby/test_groupby_shift_diff.py#L52

Removing the extraction of the data column ( `[['Z']]` ) means we now have a test that catches the bug reported in the original issue. (The test fails on master, but passes on this PR).

smithto1 committed:
#41556 passing fill_value into take_nd() allows cythonized path to be used with fill_value!=None. The existing test was extracting only the values column and ignoring that the index columns were also returned, which masked the bug reported in the issue

smithto1 committed:
Merge remote-tracking branch 'upstream/master' into issue41556

smithto1 committed:
whatsnew

smithto1 committed:
black

WillAyd commented:
Thanks for the PR. I _think_ that dispatching to `self.apply` here is going to cause a rather large performance regression though - can check the benchmarks to see?

We likely need to patch elsewhere in the groupby code to make ensure that the correct structure is returned

smithto1 commented:
> Thanks for the PR. I _think_ that dispatching to `self.apply` here is going to cause a rather large performance regression though - can check the benchmarks to see?
> 
> We likely need to patch elsewhere in the groupby code to make ensure that the correct structure is returned

I think there is a misunderstanding here. I have removed the dispatch to `self.apply` and switched it to `_get_cynthonized_result`. I haven't checked but would assume performance would improve.

simonjayhawkins commented:
> @WillAyd, I think there is a misunderstanding here. I have removed the dispatch to `self.apply` and switched it to `_get_cynthonized_result`. I haven't checked but would assume performance would improve.

does this close #26615

WillAyd reviewed (approved):
Thanks for clarifying. This lgtm

smithto1 committed:
Merge remote-tracking branch 'upstream/master' into issue41556

smithto1 commented:
@WillAyd Is there anything else we need to have this merged in?

jreback reviewed (changes requested):
needs tests that reproduce the original issues
release note needs to be update to 1.4
needs review

smithto1 committed:
whatsnew 1.4

smithto1 commented:
> needs tests that reproduce the original issues

This PR includes a modification to an existing test which ensures it recreates the original issue. The modified test fails on master (see below and attached) because of the original issue. With the code changes in this PR the test now passes. 

Previously, the test extracted the value column ( `[['Z']]` ) from the `result`, ignoring the fact that the `result` contained additional columns that should have caused an error. This is fixed now. 

```
[err1429.txt](https://github.com/pandas-dev/pandas/files/6648622/err1429.txt)

        result = g.shift(-1, fill_value=0)
    
>       tm.assert_frame_equal(result, expected)
E       AssertionError: DataFrame are different
E       
E       DataFrame shape mismatch
E       [left]:  (24, 3)
E       [right]: (24, 1)

test_groupby_shift_diff.py:60: AssertionError
```

> release note needs to be update to 1.4

Done. 

> needs review

Ready for review, @jreback

smithto1 committed:
Merge remote-tracking branch 'upstream/master' into issue41556

smithto1 commented:
@jreback can you take another look at this? I think it is ready to merge.

jreback reviewed (changes requested):
looks good. can you make sure we have an asv for this case (or add one) and show the results. ping on green. 

I think testing is ok, though if you can look in the original issue and make sure we have a replicated test.

smithto1 committed:
add ASV

pep8speaks commented:
Hello @smithto1! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`asv_bench/benchmarks/groupby.py`](https://github.com/pandas-dev/pandas/blob/5f9478f121aad55a9cbca62e321f0768cb39acbc/asv_bench/benchmarks/groupby.py):

> [Line 376:34](https://github.com/pandas-dev/pandas/blob/5f9478f121aad55a9cbca62e321f0768cb39acbc/asv_bench/benchmarks/groupby.py#L376): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 378:13](https://github.com/pandas-dev/pandas/blob/5f9478f121aad55a9cbca62e321f0768cb39acbc/asv_bench/benchmarks/groupby.py#L378): [E123](https://duckduckgo.com/?q=pep8%20E123) closing bracket does not match indentation of opening bracket's line
> [Line 379:1](https://github.com/pandas-dev/pandas/blob/5f9478f121aad55a9cbca62e321f0768cb39acbc/asv_bench/benchmarks/groupby.py#L379): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 382:1](https://github.com/pandas-dev/pandas/blob/5f9478f121aad55a9cbca62e321f0768cb39acbc/asv_bench/benchmarks/groupby.py#L382): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

smithto1 committed:
black

smithto1 commented:
> looks good. can you make sure we have an asv for this case (or add one) and show the results. ping on green.
> 
> I think testing is ok, though if you can look in the original issue and make sure we have a replicated test.

There is a replicating test. 

asv is added. Can see the difference running the same asv for master and for the PR branch. 

**master**
```
[  0.00%] ·· Benchmarking existing-pyc__anaconda3_envs_pandas-dev_python.exe
[ 25.00%] ··· groupby.Shift.time_defaults                              1.89±0ms
[ 50.00%] ··· groupby.Shift.time_fill_value                            3.47±0ms
```

**PR**
```
[  0.00%] ·· Benchmarking existing-pyc__anaconda3_envs_pandas-dev_python.exe
[ 25.00%] ··· groupby.Shift.time_defaults                              1.33±0ms
[ 50.00%] ··· groupby.Shift.time_fill_value                            1.29±0ms
```

rhshadrach commented on doc/source/whatsnew/v1.4.0.rst:
```diff
@@ -187,7 +187,7 @@ Plotting
 Groupby/resample/rolling
 ^^^^^^^^^^^^^^^^^^^^^^^^
--
+- Bug in :meth:`GroupBy.shift` that would return different columns if ``fill_value`` was not None (:issue:`41556`)
```
Can be a little more specific, different columns -> the group by columns

rhshadrach replied:
Also - can you add a line in the performance section.

smithto1 replied:
Done

rhshadrach reviewed (changes requested):
lgtm, minor request for the whatsnew

rhshadrach commented:
@smithto1 - friendly ping for resolving conflict and whatsnew request.

smithto1 committed:
whatsnew conflict

smithto1 commented:
> @smithto1 - friendly ping for resolving conflict and whatsnew request.

That is addressed. I think this is ready to merge. 

@rhshadrach @jreback @WillAyd

rhshadrach reviewed (approved):
lgtm

smithto1 committed:
fix whatsnew conflict

jreback commented on pandas/tests/groupby/test_groupby_shift_diff.py:
```diff
@@ -55,7 +55,7 @@ def test_group_shift_with_fill_value():
         index=None,
     )
-    result = g.shift(-1, fill_value=0)[["Z"]]
+    result = g.shift(-1, fill_value=0)
```
do we have sufficient coverage of the tests from the OP?

smithto1 replied:
Yes. I have investigated it. 

The original problem was that the grouping columns were improperly returned. The old form of this test would extract the value column (`[["Z"]]`), so the test did not detect that the grouping columns were also returned. I have modified this test so it no longer extracts the value column; if the grouping columns are returned (the OP) this test will now fail.

jreback replied:
kk great

if u can resolve conflicts and merge master

smithto1 replied:
@jreback conflicts resolved. I think it can be merged.

jreback reviewed (changes requested):
lgtm. test coverage question. ping on green.

smithto1 committed:
whatsnew conflict

jreback commented:
thanks @smithto1 for the patch and the patience! keep em coming!
